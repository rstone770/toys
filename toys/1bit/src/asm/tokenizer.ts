export type SourceRange = [number, number];

export interface Token<Type extends string, Value> {
  type: Type;
  value: Value;
  range: SourceRange;
}

export type AnyToken = Token<string, unknown>;

export type Tokenizer<Type extends string, Value> = (
  input: string,
  offset?: number
) => Token<Type, Value> | null;

export type TokenizerFactory = {
  <Type extends string, Value>(
    type: Type,
    pattern: RegExp,
    map: (match: RegExpMatchArray) => Value | null
  ): Tokenizer<Type, Value>;
  <Type extends string>(
    type: Type,
    pattern: RegExp,
    map?: (match: RegExpMatchArray) => string | null
  ): Tokenizer<Type, string>;
};

export type TokenizerReturnType<T extends Tokenizer<string, unknown>> = NonNullable<ReturnType<T>>;

export const tokenizer: TokenizerFactory = <Type extends string, Value>(
  type: Type,
  pattern: RegExp,
  map?: (match: RegExpMatchArray) => Value | null
): Tokenizer<Type, Value> => {
  const flags = pattern.ignoreCase ? "yi" : "y";
  const exp = new RegExp(pattern, flags);

  return (input: string, offset: number = 0) => {
    exp.lastIndex = offset;

    const match = exp.exec(input);
    if (match == null) {
      return null;
    }

    const value = map ? map(match) : match[0];
    if (value == null) {
      return null;
    }

    return {
      type,
      value: value as Value,
      range: [offset, offset + match[0].length]
    };
  };
};
