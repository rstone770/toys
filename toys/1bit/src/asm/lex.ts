import { tokenizer, TokenizerReturnType } from "./tokenizer";

export const whitespace = tokenizer("ws", /\s+/);
export const comment = tokenizer("comment", /;(.*)/, (match) => match[1]);
export const identifier = tokenizer("identifier", /[A-Z_][A-Z0-9_]*/i, (match) =>
  match[0].toLowerCase()
);

export type NumberFormat = "dec" | "hex" | "bin";
export interface Numeric {
  value: number;
  format: NumberFormat;
}
export const numeric = tokenizer(
  "numeric",
  /0x(?<hex>[0-9A-F]+)|0b(?<bin>[01]+)|[0-9]+/,
  (match): Numeric => {
    if (match.groups?.hex != null) {
      return {
        format: "hex",
        value: parseInt(match.groups.hex, 16)
      };
    }

    if (match.groups?.bin != null) {
      return {
        format: "bin",
        value: parseInt(match.groups.bin, 2)
      };
    }

    return {
      format: "dec",
      value: parseInt(match[0], 10)
    };
  }
);

export type KeywordType = "equ";
export const keyword = tokenizer(
  "keyword",
  /EQU/i,
  (match): KeywordType => match[0].toLowerCase() as KeywordType
);

export type WhitespaceToken = TokenizerReturnType<typeof whitespace>;
export type CommentToken = TokenizerReturnType<typeof comment>;
export type IdentifierToken = TokenizerReturnType<typeof identifier>;
export type NumericToken = TokenizerReturnType<typeof numeric>;
export type KeywordToken = TokenizerReturnType<typeof keyword>;

export type KnownToken =
  | WhitespaceToken
  | CommentToken
  | IdentifierToken
  | NumericToken
  | KeywordToken;

export const lex = (input: string, offset = 0): KnownToken | null => {
  if (offset >= input.length) {
    return null;
  }

  if (offset < 0) {
    return null;
  }

  return (
    whitespace(input, offset) ??
    comment(input, offset) ??
    numeric(input, offset) ??
    keyword(input, offset) ??
    identifier(input, offset)
  );
};
