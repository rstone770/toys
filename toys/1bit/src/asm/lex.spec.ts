import { expect, it, describe } from "vitest";
import { whitespace, comment, identifier, numeric, keyword, lex } from "./lex";
import { AnyToken } from "./tokenizer";

describe("whitespace", () => {
  it("matches whitespace", () => {
    expect(whitespace(" \t")).toEqual({ type: "ws", value: " \t", range: [0, 2] });
    expect(whitespace("test test", 4)).toEqual({ type: "ws", value: " ", range: [4, 5] });
  });

  it("does not match non-whitespace", () => {
    expect(whitespace("foo")).toBeNull();
    expect(whitespace("foo", 1)).toBeNull();
  });
});

describe("comment", () => {
  it("matches comments", () => {
    expect(comment("; foo")).toEqual({ type: "comment", value: " foo", range: [0, 5] });
    expect(comment("test ; foo", 5)).toEqual({ type: "comment", value: " foo", range: [5, 10] });
  });

  it("does not match non-comments", () => {
    expect(comment("foo")).toBeNull();
    expect(comment("foo", 1)).toBeNull();
  });
});

describe("identifier", () => {
  it("matches identifiers", () => {
    expect(identifier("FOO")).toEqual({ type: "identifier", value: "foo", range: [0, 3] });
    expect(identifier("foo bar", 4)).toEqual({ type: "identifier", value: "bar", range: [4, 7] });
  });

  it("does not match non-identifiers", () => {
    expect(identifier("1 foo")).toBeNull();
    expect(identifier("f1", 1)).toBeNull();
  });
});

describe("numeric", () => {
  it("matches numbers", () => {
    expect(numeric("0x64")).toEqual({
      type: "numeric",
      value: { value: 100, format: "hex" },
      range: [0, 4]
    });
    expect(numeric("0b111")).toEqual({
      type: "numeric",
      value: { value: 7, format: "bin" },
      range: [0, 5]
    });
  });

  it("does not match non-numbers", () => {
    expect(numeric("foo")).toBeNull();
    expect(numeric("foo", 1)).toBeNull();
  });
});

describe("keyword", () => {
  it("matches keywords", () => {
    expect(keyword("equ")).toEqual({ type: "keyword", value: "equ", range: [0, 3] });
  });

  it("does not match non-keywords", () => {
    expect(keyword("foo")).toBeNull();
    expect(keyword("foo", 1)).toBeNull();
  });
});

describe("lex", () => {
  const input = "foo equ 0x64 ; comment";

  it("lexes the input", () => {
    const tokens: AnyToken[] = [];
    for (let i = lex(input, 0); i != null; i = lex(input, i.range[1])) {
      tokens.push(i);
    }

    expect(tokens).toMatchSnapshot;
  });
});
