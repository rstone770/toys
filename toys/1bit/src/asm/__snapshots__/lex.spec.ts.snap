// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lex > lexes the input 1`] = `
[
  {
    "range": [
      0,
      3,
    ],
    "type": "identifier",
    "value": "foo",
  },
  {
    "range": [
      3,
      4,
    ],
    "type": "ws",
    "value": " ",
  },
  {
    "range": [
      4,
      7,
    ],
    "type": "keyword",
    "value": "equ",
  },
  {
    "range": [
      7,
      8,
    ],
    "type": "ws",
    "value": " ",
  },
  {
    "range": [
      8,
      12,
    ],
    "type": "numeric",
    "value": {
      "format": "hex",
      "value": 100,
    },
  },
  {
    "range": [
      12,
      13,
    ],
    "type": "ws",
    "value": " ",
  },
  {
    "range": [
      13,
      22,
    ],
    "type": "comment",
    "value": " comment",
  },
]
`;
